sklearn/feature_selection/univariate_selection.py

```diff
@@ -227,6 +228,188 @@ def chi2(X, y):
     return _chisquare(observed, expected)
 
 
+def _info_gain(fc_count, c_count, f_count, fc_prob, c_prob, f_prob, total):
+    def get_c_f_cell(fc_prob, c_prob, f_prob):
+        t = np.log2(fc_prob/(c_prob * f_prob))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(fc_prob, t)
+
+    def get_nc_nf_cell(fc_prob, c_prob, f_prob):
+        t = np.log2((1-f_prob-c_prob+fc_prob)/((1-c_prob)*(1-f_prob)))
+        t[~np.isfinite(t)] = 0
+        return np.multiply((1-f_prob-c_prob+fc_prob), t)
+
+    def get_c_nf_cell(c_prob, f_prob, c_count, fc_count, total):
+        nfc = (c_count - fc_count)/total
+        t = np.log2(nfc/(c_prob*(1-f_prob)))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(nfc, t)
+
+    def get_nc_f_cell(c_prob, f_prob, f_count, fc_count, total):
+        fnc = (f_count - fc_count)/total
+        t = np.log2(fnc/((1-c_prob)*f_prob))
+        t[~np.isfinite(t)] = 0
+        return np.multiply(fnc, t)
+
+    # the feature score is averaged over classes
+    return (get_c_f_cell(fc_prob, c_prob, f_prob) +
+            get_nc_nf_cell(fc_prob, c_prob, f_prob) +
+            get_c_nf_cell(c_prob, f_prob, c_count, fc_count, total) +
+            get_nc_f_cell(c_prob, f_prob, f_count, fc_count, total)).mean(
+            axis=0)
+
+
+def get_feature_class_counts(X, y):
+    """Count feature, class, joint and total frequencies
+    """
+    X = check_array(X, accept_sparse=['csr', 'coo'])
+    if np.any((X.data if issparse(X) else X) < 0):
+        raise ValueError("Input X must be non-negative.")
+
+    Y = LabelBinarizer().fit_transform(y)
+    if Y.shape[1] == 1:
+        Y = np.append(1 - Y, Y, axis=1)
+
+    # counts
+
+    fc_count = safe_sparse_dot(Y.T, X)          # n_classes * n_features
+    total = fc_count.sum(axis=0).reshape(1, -1).sum()
+    f_count = X.sum(axis=0).reshape(1, -1)
+    if issparse(X):
+        c_count = (X.sum(axis=1).reshape(1, -1) * Y).T
+    else:
+        c_count = (X.sum(axis=1) * Y.T).sum(axis=1).reshape(-1, 1)
+
+    return f_count, c_count, fc_count, total
+
+
+def get_feature_class_probs(f_count, c_count, fc_count, total):
+    """Get feature, class, joint probabilities
+    """
+    f_prob = f_count / f_count.sum()
+    c_prob = c_count / c_count.sum()
+    fc_prob = fc_count / total
+    return f_prob, c_prob, fc_prob
+
+
+def info_gain(X, y):
+    """Compute an Information Gain [1] score for each feature in the data.
+
+    The score can be used to weight features by informativeness or select the
+    most informative features for training and evaluating a classifier.
+
+    IG measures the number of bits of information obtained about the presence
+    or absence of a class by knowing the presence or absence of the feature. IG
+    of a feature is thus local to each class. This implementation outputs a
+    global IG score, by averaging the feature's scores over all classes.
```

Comment:
There is no clear advantage between  averaging, using max, or sum of class scores to get a global score for a feature. They are all valid options, see Sebastiani 2002, p. 15: http://nmis.isti.cnr.it/sebastiani/Publications/ACMCS02.pdf